bcda/models/postgres/alr.go

```diff
@@ -0,0 +1,238 @@
+		4. rollBack
+			- Used to do a rollback when a transaction fails
+*******************************************************************************/
+
+func NewAlrRepo(db *sql.DB) *AlrRepository {
+	return &AlrRepository{db}
+}
+
+func getPgxConn(db *sql.DB) *pgx.Conn {
+	conn, err := stdlib.AcquireConn(db)
+	if err != nil {
+		logrus.Warn(err, "failed to acquire connection")
+	}
+	return conn
+}
+
+func encoder(mp map[string]string) ([]byte, error) {
+	var bytea bytes.Buffer
+	enc := gob.NewEncoder(&bytea)
+	if err := enc.Encode(mp); err != nil {
+		return nil, err
+	}
+	return bytea.Bytes(), nil
+}
+
+func decoder(byteMap []byte) (map[string]string, error) {
+	var mp map[string]string
+	r := bytes.NewReader(byteMap)
+	dec := gob.NewDecoder(r)
+	if err := dec.Decode(&mp); err != nil {
+		return nil, err
+	}
+	return mp, nil
+}
+
+func rollBack(tx *pgx.Tx) {
+	if err := tx.Rollback(); err != nil {
+		logrus.Warnf("Failed to rollback transaction %s", err.Error())
+	}
+}
+
+/*******************************************************************************
+	METHODS
+		1. Next, Err, Values
+			- These methods are need to implement the CopyFrom from pgx.
+		2. AddAlr
+			- Adds new rows to the ALR Tables. A single call of this function
+			can ingest ALR data per ACO.
+		3. GetAlr
+			- Retreive data from database.
+*******************************************************************************/
+
+func (a *alrCopyFromSource) Next() bool {
+	// If we went through all of the rows, this condition should be false
+	return len(a.alrs) > 0
+}
+
+func (a *alrCopyFromSource) Err() error {
+	// Err() returns nil if "Done" has not closed and context is still valid
+	return a.ctx.Err()
+}
+
+func (a *alrCopyFromSource) Values() ([]interface{}, error) {
+	// Get the row
+	alr := a.alrs[0]
+	// Move counter up so we look at the following row next
+	a.alrs = a.alrs[1:]
+
+	// Convert the map[string]string to slices of bytes
+	bytea, err := encoder(alr.KeyValue)
+	if err != nil {
+		return nil, err
+	}
+
+	row := []interface{}{
+		&pgtype.Int8{Int: a.metaKey, Status: pgtype.Present},      // bigint
+		&pgtype.Text{String: alr.BeneMBI, Status: pgtype.Present}, // varchar
+		&pgtype.Text{String: alr.BeneHIC, Status: pgtype.Present},
+		&pgtype.Text{String: alr.BeneFirstName, Status: pgtype.Present},
+		&pgtype.Text{String: alr.BeneLastName, Status: pgtype.Present},
+		&pgtype.Text{String: alr.BeneSex, Status: pgtype.Present},
+		&pgtype.Timestamp{Time: alr.BeneDOB, Status: pgtype.Present},
+		&pgtype.Timestamp{Time: alr.BeneDOD, Status: pgtype.Present},
+		&pgtype.Bytea{Bytes: bytea, Status: pgtype.Present}, // bytea
+	}
+
+	return row, nil
+}
+
+func (r *AlrRepository) AddAlr(ctx context.Context, aco string, timestamp time.Time, alrs []models.Alr) error {
+	// Grab pgx.Conn
+	conn := getPgxConn(r.DB)
+
+	// Do this in a single transaction using BeginEX from the pgx package
+	tx, err := conn.BeginEx(ctx, nil)
+	if err != nil {
+		rollBack(tx)
```

Comment:
Since this is a little short than what is seen in import.go, the rollback done explicitly.